### PR TITLE
Implement CBOE BaseData class

### PR DESCRIPTION
**Pull request by gsalaz98**
#### Description

Adds CBOE BaseData class
#### Related Issue

N/A
#### Motivation and Context

Add additional data sources
#### Requires Documentation Change

No
#### How Has This Been Tested?

Algorithm included was used to test data
#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm.CSharp/AltData/CachedAlternativeDataAlgorithm.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            // You can use this in your algorithm as demonstrated below:
+
+            // CBOE VIX: http://cache.quantconnect.com/alternative/cboe/vix.csv
+            AddData<CBOE>("VIX");
+        }
+
+        public override void OnData(Slice data)
+        {
+            var vix = data.Get<CBOE>("VIX");
+            Log($"VIX: {Time}, {vix.Open}, {vix.High}, {vix.Low}, {vix.Close}");
```
Comment (mchandschuh) [inline Algorithm.CSharp/AltData/CachedAlternativeDataAlgorithm.cs]: Let's override the `CBOE.ToString()` in the same spirit as the other `BaseData` implementations.

File: Common/Data/Custom/CBOE/CBOE.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+using NodaTime;
+using QuantConnect.Data.Market;
+using System;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+
+namespace QuantConnect.Data.Custom.CBOE
+{
+    public class CBOE : TradeBar
```
Comment (mchandschuh) [inline Common/Data/Custom/CBOE/CBOE.cs]: Is this intended to be processed  in the same spirit as custom data? If so, I think we'll want to update the constructor to set `DataType = MarketDataType.Base`.

File: Common/Data/Custom/CBOE/CBOE.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            }
+            if (!decimal.TryParse(csv[4], NumberStyles.Any, CultureInfo.InvariantCulture, out close))
+            {
+                return null;
+            }
+
+            return new CBOE
+            {
+                // Add a day delay to the data so that we LEAN doesn't assume that we get the data at 00:00 the day of
+                Time = DateTime.Parse(csv[0], CultureInfo.InvariantCulture).AddDays(1),
```
Comment (mchandschuh) [inline Common/Data/Custom/CBOE/CBOE.cs]: Please use the helper methods added to cleanly deal with culture settings. For parsing, see the [`Parse`](https://github.com/QuantConnect/Lean/blob/master/Common/Parse.cs) class.

File: Common/Data/Custom/CBOE/CBOE.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            var csv = line.Split(',')
+                .Select(x => x.Trim())
+                .ToList();
+
+            decimal open;
+            decimal high;
+            decimal low;
+            decimal close;
+
+            if (!decimal.TryParse(csv[1], NumberStyles.Any, CultureInfo.InvariantCulture, out open))
```
Comment (mchandschuh) [inline Common/Data/Custom/CBOE/CBOE.cs]: Please consider adding a `TryParse(string, NumberStyles, out decimal)` method to the [`Parse`](https://github.com/QuantConnect/Lean/blob/master/Common/Parse.cs#L57) class. Seeing this now, I realize it was an oversight on my part to not include `TryParse` methods for the various types.

``` csharp
public static bool TryParse(string input, NumberStyles numberStyles, out decimal value)
{
    return decimal.TryParse(input, numberStyles, out value);
}
```

File: Common/Data/Custom/CBOE/CBOE.cs [hunk 5]
```diff
@@ -73,27 +86,27 @@ public override BaseData Reader(SubscriptionDataConfig config, string line, Date
             decimal low;
             decimal close;
 
-            if (!decimal.TryParse(csv[1], NumberStyles.Any, CultureInfo.InvariantCulture, out open))
+            if (!QuantConnect.Parse.TryParse(csv[1], NumberStyles.Any, out open))
```
Comment (jaredbroad) [inline Common/Data/Custom/CBOE/CBOE.cs]: Per discussion has valid values for close so shouldn't reject the whole row/point.

**Pull request merged**